designdocs/capacity_staking_rewards_implementation.md

```diff
@@ -0,0 +1,324 @@
+    /// configured Thawing Period number of blocks.
+    pub stake_change_unlocking: BoundedVec<UnlockChunk<BalanceOf<T>, T::RewardEra>, T::MaxUnlockingChunks>, // NEW
+    /// total staked amounts for each past era, up to StakingRewardsPastErasMax eras.
+    pub staking_history: BoundedVec<StakingHistory<BalanceOf<T>, T::RewardEra>, T::StakingRewardsPastErasMax>, // NEW
+}
+
+/// A per-reward-era record for StakingAccount total_staked amount.
+pub struct StakingHistory<Balance, RewardEra> {
+    total_staked: Balance,
+    reward_era: RewardEra,
+}
+```
+
+### StakingTargetDetails updates
+A new field is added to indicate the type of staking the Account holder is doing in relation to this target.
+
+```rust
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct StakingTargetDetails<T: Config> {
+	/// The total amount of tokens that have been targeted to the MSA.
+	pub amount: BalanceOf<T>,
+	/// The total Capacity that an MSA received.
+	pub capacity: BalanceOf<T>,
+	/// The type of staking, which determines ultimate capacity per staked token.
+	pub staking_type: StakingType, // NEW
+}
+```
+
+**Unstaking thaw period**
+Changes the thaw period to begin at the first block of next RewardEra instead of immediately.
+
+### Changes to extrinsics
+```rust
+pub fn stake(
+    origin: OriginFor<T>,
+    target: MessageSourceId,
+    amount: BalanceOf<T>,
+    staking_type: StakingType // NEW
+) -> DispatchResult {
+    /// NEW BEHAVIOR:
+    // if the account is new, save the new staking type
+    // if not new and staking type is different, Error::CannotChangeStakingType
+}
+
+pub fn unstake(
+    origin: OriginFor<T>,
+    target: MessageSourceId,
+    requested_amount: BalanceOf<T>,
+) -> DispatchResult {
+    // NEW BEHAVIOR:
+    // If StakingType is RewardsType
+    //   If payout_eligible,
+    //     check whether their last payout era is recent enough to pay out all rewards at once.
+    //     if so, first pay out all rewards and then continue with rest of unstaking code as is
+    //     if not, emit error "MustFirstClaimUnclaimedRewards", "UnclaimedRewardsOverTooManyEras" or something like that
+    //   If not payout eligible,
+    //     check whether the last payout era is the current one.
+    //     if so, all rewards have been claimed, so continue with rest of unstaking code as is,
+    //
+    //     otherwise, they have too many unlocking chunks so they'll have to wait. - the unstaking code
+    //     will catch this anyway and emit `MaxUnlockingChunksExceeded`
+}
+```
+### NEW: StakingRewardsProvider - Economic Model trait
+This one is not yet determined, however there are certain functions that will definitely be needed.
+The struct and method for claiming rewards is probably going to change.
+The rewards system will still need to know the `reward_pool_size`.
```

Comment:
I'm curious how are we deciding about reward pool allocation and what is the algorithm to allocate enough reward without starving or overflowing the pool?

Reply:
These decisions are above my pay grade